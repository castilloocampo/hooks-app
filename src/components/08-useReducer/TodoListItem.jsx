import React from 'react';
import { PropTypes } from 'prop-types';

const TodoListItem = ({ todo, index, onHandleDeleteTODO, onHandleToggleTODO }) => {
  const { id, description, done } = todo;
  return (
    <li
      id='list-group-item'
      className='relative block py-3 px-6 -mb-px '
      key={id}>
      <p
        className={`text-center ${done && 'line-through'}`}
        onClick={() => onHandleToggleTODO(id)} >
        {index + 1}. {description}
      </p>
      <button
        className='btn btn-danger'
        onClick={() => onHandleDeleteTODO(id)} >
        Borrar
      </button>
    </li>
  );
}

TodoListItem.propTypes = {
  todo: PropTypes.shape({
    id: PropTypes.number,
    description: PropTypes.string,
    done: PropTypes.bool
  }).isRequired,
  index: PropTypes.number.isRequired,
  onHandleDeleteTODO: PropTypes.func.isRequired,
  onHandleToggleTODO: PropTypes.func.isRequired

}

export default TodoListItem;
