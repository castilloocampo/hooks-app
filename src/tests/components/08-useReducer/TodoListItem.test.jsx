import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import TodoListItem from '../../../components/08-useReducer/TodoListItem';
import { demoTodos } from '../../fixtures/demoTodos';

describe('TodoListItem', () => {

  let fnDelete = jest.fn();
  let fnToggle = jest.fn();
  let { container, rerender } = render(<TodoListItem
    todo={demoTodos[1]}
    index={0}
    onHandleDeleteTODO={fnDelete}
    onHandleToggleTODO={fnToggle}
  />);

  beforeEach(() => {
    fnDelete = jest.fn();
    fnToggle = jest.fn();
    const tempRender = render(<TodoListItem
      todo={demoTodos[1]}
      index={0}
      onHandleDeleteTODO={fnDelete}
      onHandleToggleTODO={fnToggle}
    />);
    container = tempRender.container;
    rerender = tempRender.rerender;
  });

  afterEach(() => {
    cleanup();
  })

  test('should show correctly', () => {
    expect(container).toMatchSnapshot();
  });

  test('should call delete function', () => {
    const btnDelete = screen.getByTestId('delete-button');
    fireEvent.click(btnDelete);
    expect(fnDelete).toHaveBeenCalledTimes(1);
    expect(fnDelete).toHaveBeenCalledWith(2);
  });

  test('should call toggle function', () => {
    const paragToggle = screen.getByTestId('p-list-item');
    fireEvent.click(paragToggle);
    expect(fnToggle).toHaveBeenCalledTimes(1);
    expect(fnToggle).toHaveBeenCalledWith(2);
  });

  test('should show text correctly', () => {
    const index = 0;
    const expectedText = `${index + 1}. ${demoTodos[1].description}`;
    const paragText = screen.getByTestId('p-list-item').textContent;
    expect(paragText).toBe(expectedText);
  });

  test('should throug text', async () => {
    const todo = demoTodos[1];
    todo.done = true;
    rerender(<TodoListItem
      todo={todo}
      index={0}
      onHandleDeleteTODO={fnDelete}
      onHandleToggleTODO={fnToggle}
    />);
    const parag = screen.getByTestId('p-list-item');
    expect(parag).toHaveClass('line-through');
  });

});
