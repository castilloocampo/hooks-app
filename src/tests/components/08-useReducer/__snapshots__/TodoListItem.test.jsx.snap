// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListItem should show correctly 1`] = `
<div>
  <li
    class="relative block py-3 px-6 -mb-px "
    id="list-group-item"
  >
    <p
      class="text-center false"
      data-testid="p-list-item"
    >
      1
      . 
      Learn Mongo
    </p>
    <button
      class="btn btn-danger"
      data-testid="delete-button"
    >
      Borrar
    </button>
  </li>
</div>
`;
